### DISCUSSION TITLE
[QUESTION] Delete snapshot after successful backup ?

### DISCUSSION THREAD

**Discussion by tcoupin** [longhorn/longhorn#5797] [Q&A]
## Question

My volumes are backuped every night using velero and CSI volume snapshot with `type: bak`.
I do not need to keep snapshots in the cluster.

Is there a way to not keep snapshot for successful backup?

Currently, I'm thinking to add cronjob to execute:
```
kubectl get backups.longhorn.io -n longhorn-system -o custom-columns=SNAPSHOT:.spec.snapshotName | grep '^snapshot-'| xargs kubectl delete snapshots.longhorn.io -n longhorn-system
```

## Environment

 - Longhorn version: 1.4.0
 - Kubernetes version: 1.24
 - Node config
   - OS type and version ubuntu 22.04
   - CPU per node: 4
   - Memory per node: 8
   - Disk type hdd
   - Network bandwidth and latency between the nodes: 10G
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): vmware, rke2

**Comment by weizhe0422** [comment]
I think there is an optional parameter: `--snapshot-volumes=false` to disable creating snapshots when doing Velero backup creation.

ref. https://velero.io/docs/v1.4/how-velero-works/

**Reply by R-Studio** [reply]
If you change this, then Velero don't trigger a backup of the persistent volumes.

**Comment by Pro** [comment]
@tcoupin did you already find a good solution for your question?

I have the exact same use-case. I.e., using `type: bak` with velero and storing the backups in S3 backend.
So I am guessing that the snapshots within Longhorn are actually not needed and could be deleted to save space.

The big question is: will restore actions with velero still work.

**Reply by tcoupin** [reply]
No, I have scheduled a cronjob to delete useless snapshots (longhorn.io/v1beta2). Velero works with volumesnapshots (snapshot.storage.k8s.io/v1) and the created volumesnapshotcontents is linked to the backups (longhorn.io/v1beta2) so there is no problem to restore pvc.

**Comment by ChanYiLin** [comment]
Hi @tcoupin @Pro 
Thanks for choosing Longhorn as your storage solution.
We plan to introduce this feature in this feature request: https://github.com/longhorn/longhorn/issues/9213
Please stay tuned for further updates.
And feel free to comment in this ticket.
Thanks

**Reply by ChanYiLin** [reply]
Before this feature is implemented.
You can also rely on our recurring job to periodically clean up snapshots and maintain the retention limit.
https://longhorn.io/docs/1.7.2/snapshots-and-backups/scheduling-backups-and-snapshots/#set-up-recurring-jobs
Thanks

**Reply by vanchaxy** [reply]
@ChanYiLin Hi, I'm not sure the feature request is directly related to this issue.

I'm experiencing a similar problem, but I'm using VolSync instead of Velero. The issue arises when VolSync clones the volume into a temporary PVC for backup. Longhorn creates a new snapshot for every clone, so after 250 backups, I’m hitting the snapshot limit.

The linked feature request addresses backups created by Longhorn itself, but I believe the problem will still exist for other solutions like VolSync.

I'm using the workaround you mentioned, but it would be great to have a built-in way to clean up only the "cloned" snapshots. These snapshots already have the label longhorn.io/for-cloning-volume, so it's possible to delete them with an external script, but having a native solution would be much nicer.
